ADD: Document migrating existing clusters to use groups. (#698)
Kuadrant/dns-operator: docs/migrating_existing_clusters_to_use_groups.md

# Migrating Existing Clusters To Use Groups

## Why use Groups

> By using groups it allows for planned migrations, and failovers in unexpected outages of deployments.

Reviewer: ```suggestion
Using groups allows for planned migrations, and for fail-overs when unexpected outages occur.
```

- providerRef, this is the secret used by the dns-operator to connect to the provider. Format required \<namespace\>/\<name\>.
- domain, this is the route domain for the zone.
With this information the CLI command to use is.
```sh
kubectl-kuadrant_dns get-active-groups --providerRef <namespace>/<name> --domain <domain>
```

#### Manually via the DNS provider
In the DNS provider there will be a TXT record created that states the active groups.
This TXT record has a naming schema of `kuadrant-active-groups.<domain>`.
The active groups are in a `&&` separated list.

Below is a sample TXT record.

```txt
Record name: kuadrant-active-groups.<domain>
Record type: TXT
Value: "version=1;groups=GROUP_ID1&&GROUP_ID2"
```

The TXT record can also be accessed by doing a dig on the TXT record.

```sh
dig kuadrant-active-groups.<domain> TXT +short
```

### Configuring the dns-operator to be part of a group
The dns-operator can be added to a group via an env var set in the `dns-operator-controller-env` configmap.
This configmap exists in the same namespace as the dns-operator.

By using kubectl patch, the configmap can be updated with affecting other env vars that may be configured.
```sh

> kubectl patch configmap dns-operator-controller-env --namespace <namespace> --type merge --patch '{"data:{"GROUP":"<GROUP ID>"}}'

Reviewer 2: Is there a typo here? `"data:` -> `"data":`

```sh
kubectl rollout restart deployment dns-operator-controller-manager --namespace <namespace>
```

### Confirming the dns-operator is configured correctly

It can take some time for the all the dnsrecords to be fully reconciled to use the new group setting.
However, by checking the logs of the operator, it can be confirmed that the group was configured correctly.
There is are two setup log messages that confirm the group was correctly configured.
These log messages can be retrieved from the cluster with the following command.
```sh
NS=<namespace> kubectl logs $(kubectl get pods -l control-plane=dns-operator-controller-manager --sort-by=.metadata.creationTimestamp -o name --namespace $NS | tail -n 1) --namespace $NS | head -n 20 | grep group

# expected output
{"level":"info","ts":"2006-1-2T15:4:5","logger":"setup","msg":"overriding group flag with \"group1\" value"}
{"level":"info","ts":"2006-1-2T15:4:5","logger":"setup","msg":"using group: group1"}
```

When the reconcile finally completes on the dnsrecords in the status block there will be mention of the group.
The current group of the dnsrecords can be checked with the following command.
```sh

> kubectl get dnsrecords -A -o custom-columns="NAMESPACE:.metadata.namespace,NAME:.metadata.name,OWNER_ID:.status.ownerID,STATUS.GROUP:.status.group"

Reviewer 3: ```suggestion
kubectl get dnsrecords.kuadrant.io -A -o custom-columns="NAMESPACE:.metadata.namespace,NAME:.metadata.name,OWNER_ID:.status.ownerID,STATUS.GROUP:.status.group"
```

openshift has other default dnsrecords CR

Reviewer: let's make a ticket on the backlog to track this change, and link it here.
